builtin/vector_test.go

```diff
@@ -0,0 +1,118 @@
+package builtin
+
+import (
+	"testing"
+
+	"github.com/spy16/slurp/core"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestVectorIsHashable(t *testing.T) {
+	t.Parallel()
+	defer func() {
+		if r := recover(); r != nil {
+			t.Error("PersistentVector is not hashable.")
+		}
+	}()
+
+	m := make(map[core.Vector]struct{})
+	m[EmptyVector] = struct{}{}
```

Comment:
If we go with the `Hash() ([]byte, error)` model that we considered in Sabre, we would not need to make the type itself hash-able by Go runtime right? 🤔 

vector can provide a `Hash() ([]byte, error)` method that hash-based data structures make use of to generate the hash.. (internally vector can use any approach to do that -- but i am not sure what approaches are possible)..

Reply:
> If we go with the Hash() ([]byte, error) model that we considered in Sabre, we would not need to make the type itself hash-able by Go runtime right? 🤔

Correct.  I actually think our mental models for `Hashable` diverged a while back, though. Let me explain. 😄 

I had initially proposed the following interface for `Hashable`:

```go
type interface Hashable {
    Hash() interface{}  // should probably be called Hashable(), actually...
}
```

The intent was for `Hash()` to provide a hashable representation of the value, i.e. one that can be consumed by the Go runtime's hash function (`builtin.nilinterhash`).  While `[]byte` is certainly a "hashable representation", there's a subtle distinction:  I wanted `Hash()` to be implemented _only_ for types that were not natively hashable in Go.  In other words, I think the data-structures in `builtin` should use the Go runtime's built-in hash function!

Here's an illustration of how the pieces fit together.  I think code is clearer than text, here!  https://play.golang.org/p/0-QsKSmQNWg

I think the confusion was introduced when I said that `Hashable` decouples the hashability of a type from the actual hash algorithm.  While the `Hashable` interface does indeed leave the choice of hash algorithm to the caller, I did **not** mean to imply we should maintain direct compatibility with [`hash.Hash`](https://golang.org/pkg/hash/#Hash)!  The conversion from `interface{}` to `[]byte` should still be left as an exercise to the user, either through serialization or via [something more "clever"](https://play.golang.org/p/TVJyJQVzgRy).  Note that this is how Go's `hash` library is designed as well -- it is the user's responsibility to marshal their data-structures into `[]byte` prior to hashing.

I therefore think it's fine -- and in fact even *preferable* -- for Slurp's internals to be tightly coupled to the Go runtime.  That is to say, the types in the `builtin` package should use Go's `nilinterhash` because it's going to be the most optimized, standardized hash function at our disposal.  My point about decoupling was simply that the `Hashable` interface avoids  imposing `nilinterhash` on _others_.  External libraries can hash however they please.

Returning to your question, I hope it now makes sense why I would prefer `PersistentVector` to be natively hashable!

Reply:
Ahh Right, that makes sense 😅. Thanks for clarifying.

Reply:
No worries!  I don't think I was very clear the first time around. 😄 

I'm closing this conversation for now, but feel free to re-open.